vowpalwabbit/marginal.cc

```diff
@@ -15,82 +17,110 @@ namespace MARGINAL
 {
 struct expert
 {
-  float regret;
-  float abs_regret;
-  float weight;
+  expert() = default;
+  expert(float regret, float abs_regret, float weight) : regret(regret), abs_regret(abs_regret), weight(weight) {}
+
+  float regret = 0.f;
+  float abs_regret = 0.f;
+  float weight = 1.f;
 };
 
 typedef std::pair<double, double> marginal;
 typedef std::pair<expert, expert> expert_pair;
 
 struct data
 {
+  data(float initial_numerator, float initial_denominator, float decay, bool update_before_learn,
+      bool unweighted_marginals, bool compete, parameters* m_weights, loss_function* m_loss_function,
+      shared_data* m_shared_data)
+      : initial_numerator(initial_numerator)
+      , initial_denominator(initial_denominator)
+      , decay(decay)
+      , update_before_learn(update_before_learn)
+      , unweighted_marginals(unweighted_marginals)
+      , compete(compete)
+      , m_weights(m_weights)
+      , m_loss_function(m_loss_function)
+      , m_shared_data(m_shared_data)
+  {
+    id_features.fill(false);
+  }
+
+  data(float initial_numerator, float initial_denominator, float decay, bool update_before_learn,
```

Comment:
i am guessing `all` will never be `nullptr` right?

Reply:
It should never be in the way it is currently used. Especially since it is not stored I can change it to a reference